### PR TITLE
[#28] Updated for changes introduced in 4.2.9 and added new rule (master)

### PR DESCRIPTION
**Pull request by korydraughn**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -79,8 +79,30 @@ should be similar to the following:
...
+- pep_api_data_obj_rename_pre
+- pep_api_data_obj_unlink_ppost
+- pep_api_data_obj_unlink_pre
+- pep_api_replica_close_post
+- pep_api_replica_close_pre
+- pep_api_replica_open_pre
+- pep_api_rm_coll_post
+- pep_api_rm_coll_pre
 
 Even though this plugin will process PEPs first due to it's positioning, subsequent Rule Engine Plugins (REP) will
```
Comment (trel) [inline README.md:107]: "it's" -> "its"

File: README.md [hunk 2]
```diff
@@ -79,8 +79,30 @@ should be similar to the following:
 ```
 
 ## Configuration
-To enable, prepend the following plugin config to the list of rule engines in `/etc/irods/server_config.json`. 
-The plugin config must be placed ahead of all plugins that do not support continuation.
+To enable, prepend the following plugin configuration to the list of rule engines in `/etc/irods/server_config.json`. 
+
+The plugin configuration must be placed ahead of all plugins that define the following PEPs:
```
Comment (trel) [inline README.md:84]: "that define the following"
to
"that define any of the following"

File: README.md [hunk 3]
```diff
@@ -151,9 +174,36 @@ Logical Quotas only provides a relative value assuming there are many clients ac
...
+The **keys** are derived from the **namespace** and **metadata_attribute_names** defined by the plugin configuration.
+
 ### Invoking operations via the Native Rule Language
 Here, we demonstrate how to start monitoring a collection just like in the section above.
 ```bash
 $ irule -r irods_rule_engine_plugin-irods_rule_language-instance 'logical_quotas_start_monitoring_collection(*col)' '*col=/tempZone/home/rods' ruleExecOut
 ```
 
+## Stream Operations
+With previous interations of the plugin, changes in data were tracked and checked for violations across all
```
Comment (trel) [inline README.md:199]: "interations" -> "iterations"

"the plugin" -> "this plugin"

File: README.md [hunk 4]
```diff
@@ -151,9 +174,36 @@ Logical Quotas only provides a relative value assuming there are many clients ac
...
+stream-based operations in real-time. However, with the introduction of intermediate replicas and logical-locking,
```
Comment (trel) [inline README.md:200]: "logical-locking" -> "logical locking in iRODS 4.2.9"

File: README.md [hunk 5]
```diff
@@ -151,9 +174,36 @@ Logical Quotas only provides a relative value assuming there are many clients ac
...
+maintaining this behavior became complex. Due to the complexity, the handling of quotas have been relaxed. The most
```
Comment (trel) [inline README.md:201]: "have" -> "has"

File: src/handler.cpp [hunk 6]
```diff
@@ -479,17 +464,116 @@ namespace
...
+
+                    auto* out = static_cast<ExecCmdOut*>(msp->inOutStruct);
+                    std::memset(out, 0, sizeof(ExecCmdOut));
+
+                    // Copy the JSON string into the output object.
+                    const auto json_string = quota_status.dump();
+                    const auto buffer_size = json_string.size() + 1;
+                    out->stdoutBuf.len = buffer_size;
+                    out->stdoutBuf.buf = std::malloc(sizeof(char) * buffer_size);
+                    std::memcpy(out->stdoutBuf.buf, json_string.data(), buffer_size);
```
Comment (trel) [inline src/handler.cpp:531]: regarding the + 1 up there... this handles a null at the end of the buffer cleanly?

File: README.md [hunk 7]
```diff
@@ -112,6 +134,7 @@ The following operations are supported:
 - logical_quotas_set_maximum_number_of_data_objects
 - logical_quotas_set_maximum_size_in_bytes
 - logical_quotas_start_monitoring_collection
+- logical_quotas_status
```
Comment (trel) [inline README.md:137]: i'm realizing all the other policies start with a verb....

how do you feel about "logical_quotas_get_collection_status"?

File: src/handler.cpp (hunk 6 above)
Comment (korydraughn) [reply src/handler.cpp:531]: Yes. The size function does not include the null terminator, so we add 1 to it for the null byte.

File: README.md (hunk 7 above)
Comment (korydraughn) [reply README.md]: Sounds good.

**Comment by korydraughn** [comment]
This PR relaxes the handling of quotas during stream-based operations. In previous iterations of this plugin, seeks and writes were tracked and checked in real-time. With iRODS v4.2.9 comes intermediate replicas and logical locking. These two enhancements increase the complexity around tracking stream-based writes and seeks. For that reason, this PR removes the real-time tracking of writes and seeks, and instead, enforces quotas when opening and closing data objects. This ultimately means that stream-based writes are free to violate the maximum bytes quota if the open is successful. However, once the quota is violated, new calls to open will fail immediately.

@anderbubble Do you see any issues with solution that may impact you or others negatively?

**Comment by trel** [comment]
looks good to me - # when ready

**Comment by korydraughn** [comment]
Added #

**Pull request merged**